[css-view-transitions] Implement 'auto' value for view-transition-name.

mattwoodrow opened the pull request:
#### 5aed448416d432a624ec51b3a77fafc418f6683d
<pre>
[css-view-transitions] Implement &apos;auto&apos; value for view-transition-name.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281203">https://bugs.webkit.org/show_bug.cgi?id=281203</a>
&lt;<a href="https://rdar.apple.com/137656730">rdar://137656730</a>&gt;

Reviewed by Tim Nguyen.

Support the &apos;auto&apos; value by using the Element&apos;s ID if available, otherwise a generated string.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/parsing/view-transition-name-valid-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.cpp:
(WebCore::CSSPropertyParserHelpers::consumeViewTransitionName): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionName):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::isCrossDocument):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::hasViewTransitionName const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialViewTransitionName):
(WebCore::RenderStyle::viewTransitionName const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setViewTransitionName):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/ViewTransitionName.h: Added.
(WebCore::ViewTransitionName::createWithNone):
(WebCore::ViewTransitionName::createWithAuto):
(WebCore::ViewTransitionName::createWithCustomIdent):
(WebCore::ViewTransitionName::isNone const):
(WebCore::ViewTransitionName::isAuto const):
(WebCore::ViewTransitionName::isCustomIdent const):
(WebCore::ViewTransitionName::customIdent const):
(WebCore::ViewTransitionName::scopeOrdinal const):
(WebCore::ViewTransitionName::ViewTransitionName):
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertViewTransitionName):

Canonical link: <a href="https://commits.webkit.org/285041@main">https://commits.webkit.org/285041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41e65146783a4aa5cfe24f8cb254a0403d359684

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71296 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24069 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75404 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22501 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22320 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56362 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46063 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61443 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36798 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18904 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20842 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64629 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77126 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15530 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18455 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15573 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64058 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15783 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12197 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5831 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46509 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1288 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47580 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48863 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/style/ViewTransitionName.h:
```diff
@@ -0,0 +1,114 @@
+
+#include "StyleScopeOrdinal.h"
+
+namespace WebCore {
```
Can we use `WebCore::Style` ? For new types we like to namespace with Style if possible (e.g. Style::ScopedName, but also Sam's recent refactorings go in that direction)

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -459,12 +470,14 @@ ExceptionOr<void> ViewTransition::captureOldState()
 
                 // FIXME: Skip fragmented content.
 
+                ALWAYS_LOG_WITH_STREAM(stream << "CaptureOldState with name " << name);
```
Leftover ?

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -509,10 +523,12 @@ ExceptionOr<void> ViewTransition::captureNewState()
                 if (auto check = checkDuplicateViewTransitionName(name, usedTransitionNames); check.hasException())
                     return check.releaseException();
 
+                ALWAYS_LOG_WITH_STREAM(stream << "captureNewState with name " << name);
```
Leftover?

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -487,8 +500,9 @@ ExceptionOr<void> ViewTransition::captureOldState()
-        auto transitionName = renderer->style().viewTransitionName();
-        m_namedElements.add(transitionName->name, capture);
+        auto transitionName = effectiveViewTransitionName(renderer, styleable->element, document()->styleScope(), isCrossDocument());
+        ALWAYS_LOG_WITH_STREAM(stream << "CaptureOldState 2 with name " << transitionName);
```
Leftover?

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -215,6 +215,8 @@ class ViewTransition : public RefCounted<ViewTransition>, public VisibilityChang
     // ActiveDOMObject.
     void stop() final;
 
+    bool isCrossDocument() { return !m_shouldCallUpdateCallback; }
```
Can we rename the member as well? and change the usage in `callUpdateCallback()` ? 

I find this member name a bit confusing, since you could easily confuse it for the case no update callback is provided.

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -313,17 +313,28 @@ void ViewTransition::setupViewTransition()
+    StringBuilder builder;
+    builder.append("-ua-auto-"_s);
+    builder.append(String::number(originatingElement.identifier().toRawValue()));
+    return builder.toAtomString();
```
It's probably more performant and shorter to use `makeAtomString` here:
```
return makeAtomString("-ua-auto-"_s, String::number(originatingElement.identifier().toRawValue()));

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -313,17 +313,28 @@ void ViewTransition::setupViewTransition()
+        return transitionName.customIdent();
+    ASSERT(transitionName.isAuto());
+    if (originatingElement.hasID())
+        return originatingElement.getIdAttribute();
```
`originatingElement` actually isn't right in the context of pseudo-elements, you'll actually want `renderer.element()` here. In the case of pseudo-elements, you don't want `#foo::after` to transition with `#foo`.

Maybe worth a WPT and a comment.

nt1m commented on Source/WebCore/rendering/style/ViewTransitionName.h:
```diff
@@ -0,0 +1,114 @@
+    if (name.isAuto())
+        ts << "auto";
+    else if (name.isNone())
+        ts << "none";
```
```suggestion
        ts << "auto"_s;
    else if (name.isNone())
        ts << "none"_s;
```

mattwoodrow committed:
[css-view-transitions] Implement 'auto' value for view-transition-name.
https://bugs.webkit.org/show_bug.cgi?id=281203
<rdar://137656730>

Reviewed by Tim Nguyen.

Support the 'auto' value by using the Element's ID if available, otherwise a generated string.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/parsing/view-transition-name-valid-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.cpp:
(WebCore::CSSPropertyParserHelpers::consumeViewTransitionName): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionName):
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::captureNewState):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::isCrossDocument):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::hasViewTransitionName const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialViewTransitionName):
(WebCore::RenderStyle::viewTransitionName const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setViewTransitionName):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/ViewTransitionName.h: Added.
(WebCore::ViewTransitionName::createWithNone):
(WebCore::ViewTransitionName::createWithAuto):
(WebCore::ViewTransitionName::createWithCustomIdent):
(WebCore::ViewTransitionName::isNone const):
(WebCore::ViewTransitionName::isAuto const):
(WebCore::ViewTransitionName::isCustomIdent const):
(WebCore::ViewTransitionName::customIdent const):
(WebCore::ViewTransitionName::scopeOrdinal const):
(WebCore::ViewTransitionName::ViewTransitionName):
(WebCore::operator<<):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertViewTransitionName):

Canonical link: https://commits.webkit.org/285041@main
